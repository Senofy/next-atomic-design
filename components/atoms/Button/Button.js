import styled from 'styled-components'
import colors from '~/globalColors'

export const Button = styled.button`
	font-size: 40px;
	color: ${colors['blue']['300']};
	border-radius: 3px;
`
