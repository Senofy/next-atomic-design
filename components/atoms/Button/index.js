import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'
import { font, palette } from 'styled-theme'

// Functions to turn props into usable css
const backgroundColor = ({ backgroundColor, backgroundColorWeight }) => palette(backgroundColor || 'primary', backgroundColorWeight || 0)
const color = ({ color, colorWeight }) => palette(color || 'white', colorWeight || 0)

// Paragraph
export const Button = styled.button`
    font-family: ${font('primary')};
    font-size: ${({ fontSize }) => fontSize ? `${fontSize}px` : '12px'};
    font-weight: ${({ fontWeight }) => fontWeight};
    line-height: ${({ lineHeight }) => lineHeight};
    background-color: ${backgroundColor};
    color: ${color};

    padding: 0.46875rem 0.9375rem;

    border-radius: 0.25rem;
`

Button.propTypes = {
    reverse: PropTypes.bool,
    fontSize: PropTypes.number,
    lineHeight: PropTypes.number
}

Button.defaultProps = {
    palette: 'grayscale',
    fontSize: 12
}