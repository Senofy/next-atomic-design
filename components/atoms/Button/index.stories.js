import { storiesOf } from '@storybook/react'
import { P } from './'

storiesOf('Atom', module).add('default', () => <P></P>)
