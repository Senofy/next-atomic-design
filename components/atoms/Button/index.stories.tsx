import React from 'react'
import { Story, Meta } from '@storybook/react'

import { Button, ButtonProps } from '.'

export default {
	title: 'atoms/Button',
	component: Button,
	argTypes: {
		children: { control: 'text' },
	},
	args: {
		children: 'Button',
	},
} as Meta

const Template: Story<ButtonProps> = (args) => <Button {...args} />

export const Normal = Template.bind({})
