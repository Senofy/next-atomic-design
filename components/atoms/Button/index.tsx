import styled from 'styled-components'
import {
	alignSelf,
	AlignSelfProps,
	borderRadius,
	BorderRadiusProps,
	color,
	ColorProps,
	fontFamily,
	FontFamilyProps,
	fontSize,
	FontSizeProps,
	fontWeight,
	FontWeightProps,
	height,
	HeightProps,
	justifySelf,
	JustifySelfProps,
	lineHeight,
	margin,
	MarginProps,
	padding,
	PaddingProps,
	width,
	WidthProps,
} from 'styled-system'

export interface ButtonProps
	extends ColorProps,
		FontFamilyProps,
		FontSizeProps,
		FontWeightProps,
		PaddingProps,
		BorderRadiusProps,
		WidthProps,
		HeightProps,
		MarginProps,
		AlignSelfProps,
		JustifySelfProps {
	darkenOnHover?: boolean
	textDecoration?: string
}

export const Button = styled.button<ButtonProps>`
	${color};
	${fontFamily};
	${fontSize};
	${fontWeight};
	${lineHeight};
	${padding};
	${margin};
	${alignSelf};
	${justifySelf};

	text-decoration: ${({ textDecoration }) => textDecoration};

	height: max-content;
	${width};
	${height};

	border-radius: 4px;
	${borderRadius};

	display: flex;
	justify-content: center;
	align-items: center;

	transition: color 0.15s ease-in-out, background-color 0.15s ease-in-out, border-color 0.15s ease-in-out,
		box-shadow 0.15s ease-in-out;

	&:hover {
		box-shadow: ${({ darkenOnHover }) => darkenOnHover && 'inset 0px 0px 400px 110px rgba(0, 0, 0, 0.1)'};
	}
`

Button.defaultProps = {
	color: 'white',
	backgroundColor: 'green-500',
	fontFamily: 'body',
	fontSize: 14,
	padding: '8px 12px',
	darkenOnHover: true,
}
