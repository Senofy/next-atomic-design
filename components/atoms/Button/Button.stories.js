/* @flow */

import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import { Button } from './Button.js'

storiesOf('Button', module).add('default', () => <Button onClick={action('clicked')}>Learn more</Button>)
