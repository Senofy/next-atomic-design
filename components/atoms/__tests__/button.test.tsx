import React from 'react'
import { render, cleanup, fireEvent } from '@testing-library/react'
import { Button } from '../Button'

afterEach(cleanup)

test('<Button> should render without error', () => {
	const { asFragment } = render(<Button>CLick Me</Button>)
	expect(asFragment()).toMatchSnapshot()
})

test('<Button> renders children when passed in', () => {
	const { getByText } = render(<Button>Hello</Button>)
	expect(getByText(/hello/i)).toBeInTheDocument()
})

test('<Button> calls correct function on click', () => {
	const onClick = jest.fn()
	const { getByText } = render(<Button onClick={onClick}>Submit</Button>)
	const button = getByText(/submit/i)
	fireEvent.click(button)
	expect(onClick).toHaveBeenCalled()
})
