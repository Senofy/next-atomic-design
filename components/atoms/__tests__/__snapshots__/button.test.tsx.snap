// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> should render without error 1`] = `
<DocumentFragment>
  <button
    class="Button-sc-1s4wg5s-0 gvwdPA"
    color="white"
    font-family="body"
    font-size="14"
  >
    CLick Me
  </button>
</DocumentFragment>
`;
